### DISCUSSION TITLE
ExtPay API Errors

### DISCUSSION THREAD

**Discussion by nobletime** [Glench/ExtPay#62] [Q&A] [answered]
I feel the code is buggy and hasn't been tested, it throws error in fetching the user, so both samples fail miserably for me. The fetch returns 404 not found. Anyone has had any luck with ExtPay? 

	    async function fetch_user() {
	        var storage = await get(['extensionpay_user', 'extensionpay_installed_at']);
	        const api_key = await get_key();
	        if (!api_key) {
	            return {
	                paid: false,
	                paidAt: null,
	                installedAt: storage.extensionpay_installed_at ? new Date(storage.extensionpay_installed_at) : new Date(), // sometimes this function gets called before the initial install time can be flushed to storage
	                trialStartedAt: null,
	            }
	        }

	        const resp = await fetch(`${EXTENSION_URL}/api/user?api_key=${api_key}`, {
	            method: 'GET',
	            headers: {
	                'Accept': 'application/json',
	            }
	        });

**Comment by Glench** [comment, marked as answer]
Hello friend! Can you please provide steps and code to reproduce the issue you are getting? ExtPay is working great for hundreds of developers in production so it may just be an issue with your particular code. Or maybe you found a bug!

I also suggest you uninstall the extension in your browser and reinstall it. Sometimes a 404 can happen if you change your extension's ExtPay id (since there's a mismatch between the extension user's API key and the extension it's used for).

**Comment by nobletime** [comment]
That actually worked! I'm not sure why the first time I changed the ID and loaded folder unpack it didn't work. Thank you for your reply. I'm happy to see it working, thanks again!

**Comment by nobletime** [comment]
I had a chance to try it out and must say it works great. One thing I found that if a trial user has already started out, once extension is removed and re-installed it wont recognize the trial user and the login doesn't recognize trial user either, as only paid users can login. Is there a way to have an active trial user continue his trial after re-installing the extension?

extpay.onTrialStarted.addListener call back didn't work for me, is this event triggered after user clicks on confirmation email from setting up their trial?

Thanks!

**Reply by Glench** [reply]
There is no way for a trial user to restart their trial. If a user actually runs into that in practice please email me and we can probably figure something out.

onTrialStarted only fires after the user clicks the confirmation email link and it opens in the browser that they have the extension installed.

**Reply by nobletime** [reply]
Ok thank you for your reply!  That's fine!

Separately, for subscriptions is there a way to notify user of the next billing cycle ? or the subscription period, when the next due date is? 

since PaidAt seemingly only shows the first paid date, I can not do internally calculate the billing cycle by adding a month to it as active subscription cycle. Can that value be added to user.subscription?

**Reply by Glench** [reply]
You could pretty easily just take the paidAt datetime and figure out the next billing date from that.

**Comment by nobletime** [comment]
Thank you! 

Would PaidAt show the most recent date of payment in "reoccurring subscription model"?

How can I delete an extension ID? It lets me edit one but i dont see delete

**Comment by Glench** [comment]
It would show the first date of payment. There's no way to delete extension IDs, just rename it something like (old).

**Reply by nobletime** [reply]
Renaming wouldn't work because it won't allow for example one time payment to be changed to monthly subscriptions and vice versa. 

First date of payment also wouldn't help to know the active subscription month, it would be complicated as we have to know how many months passed since the first date of payment and add to it. Why not simply return the active month from stripe?

**Reply by Glench** [reply]
I mean, if you don't have any actual users yet, create a new extension ID and rename the old one something like (old) or (deleted).

And yeah, I have other priorities at the moment so I won't be doing that. It's a good suggestion but in the meantime I recommend just grabbing the date of subscription (e.g. the 29th) and checking when the next one is.

**Reply by nobletime** [reply]
Ok, well then why is there a forum or support to get feedback if this product is not actively being maintained due to lack of priority? Lol

Anyhow, it's all good! 

I also noticed that if trial is enabled on the extension, one can keep uninstalling and installing and using different email to signup with new trial to continue using the product. It's easy for people to make or provide a new email and get a trial version.

**Reply by Glench** [reply]
Hey man, look, I understand you're trying to get your stuff done and feel frustrated, but I'd appreciate it if you would be less critical. ExtensionPay is a complicated piece of software and there are many parts to running a business. I've been very responsive in supporting you so far even if I didn't add the one feature you wanted. Please try to be a bit more understanding.

Even so, I hear you've found that people can create new email addresses to get many trials. Yep! This is true of all trial-based software that uses email addresses. I haven't seen anyone really abuse this so far. One thing that can help is making the trial short enough so that it's annoying for users to constantly have to keep uninstalling/reinstalling/reactivating their extension.

**Reply by burkybang** [reply]
@nobletime Create a new discussion post specifically requesting your desired feature. Be specific. If enough people upvote it, @Glench can classify it as a higher priority. That's one of the purposes of these discussion forums.

**Reply by nobletime** [reply]
Sounds good! 

Will extensionpay support PayPal? The website says coming soon, is there a timeline?

**Reply by burkybang** [reply]
I can't speak as to when since I'm not Glen, but you can upvote this discussion post to show your interest: https://github.com/Glench/ExtPay/discussions/53
